### ISSUE TITLE
pihole-FTL: No process found

### ISSUE THREAD

**Issue by dracenmarx** [pi-hole/pi-hole#1809]
**In raising this issue, I confirm the following:** `{please fill the checkboxes, e.g: [X]}`

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [x] The issue I am reporting can be *replicated*.
- [x] The issue I am reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the the source code relevant to this issue?:**

4

---

During installation, I get following error message:
> [â] Force-reloading DNS service
> pihole-FTL: No process found

/etc/.pihole/pihole has at line 363:
```
  # Send signal to FTL to have it re-parse the gravity files
  killall -s SIGHUP pihole-FTL
```

Indeed, this does not work on my system:

> $ killall -s SIGHUP pihole-FTL
> pihole-FTL: No process found

However, this works:
> $ service pihole-FTL stop
> $ service pihole-FTL start

I already tried an update and a re-configuration.

**My system:**

> $ uname -a
> Linux ras3 4.9.66-v7+ #1059 SMP Fri Dec 1 15:38:42 GMT 2017 armv7l GNU/Linux
> $ pihole -d
> [â] Your debug token is: ptgfu86t6y

**Comment by Mcat12** [comment]
What is the output of these commands?
```
which pihole-FTL
file /usr/bin/pihole-FTL
```

**Comment by dracenmarx** [comment]
The output is 

```
root@pi:~$ which pihole-FTL
/usr/bin/pihole-FTL
root@pi:~$ file /usr/bin/pihole-FTL
/usr/bin/pihole-FTL: ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-armhf.so.3, for GNU/Linux 2.6.32, BuildID[sha1]=3eeb36a66514a98d2e10e3b673e6e93fda0f1380, not stripped
```

Also, I found out that pihole-FTL was not running, and it can be started without problems:

```
root@pi:~$ ps aux | grep pihole
root     30755  0.0  0.1   4740  1664 pts/1    S+   08:37   0:00 grep pihole
root@pi:~$ pihole-FTL
FTL started!
root@pi:~$ killall -s SIGHUP pihole-FTL
root@pi:~$ pihole-FTL
Found pihole-FTL process with PID 30785 (my PID 30876) - killing it ...
FTL started!
```

**Comment by Mcat12** [comment]
You should have FTL started through systemd instead of under root, because it will mess up file permissions otherwise (it should be using the `pihole` user).

**Comment by dracenmarx** [comment]
I only ran it manually once because I wanted to find out if there is a problem. 
The current status is, that this process does not run out-of-the-box (With normal install/repair procedures) on my system. What can I do?

**Comment by Mcat12** [comment]
If you start it via `sudo service pihole-FTL start` then do a repair, does it work?

**Comment by dracenmarx** [comment]
@Mcat12 When I start the service and do a repair (or simply call /opt/pihole/gravity.sh ), then the message still shows up.

Also, I found out that the status of the service is "active (exited)". Does that mean that it shuts itself down after starting?

```
root@ras3:~$ sudo service pihole-FTL stop
root@ras3:~$ sudo service pihole-FTL start
root@ras3:~$ sudo service pihole-FTL status
* pihole-FTL.service - LSB: pihole-FTL daemon
   Loaded: loaded (/etc/init.d/pihole-FTL; generated; vendor preset: enabled)
   Active: active (exited) since Sun 2017-12-10 11:46:20 CET; 5s ago
     Docs: man:systemd-sysv-generator(8)
  Process: 14554 ExecStop=/etc/init.d/pihole-FTL stop (code=exited, status=0/SUCCESS)
  Process: 14656 ExecStart=/etc/init.d/pihole-FTL start (code=exited, status=0/SUCCESS)
   CGroup: /system.slice/pihole-FTL.service

Dez 10 11:46:04 ras3 systemd[1]: Starting LSB: pihole-FTL daemon...
Dez 10 11:46:04 ras3 pihole-FTL[14656]: Not running
Dez 10 11:46:04 ras3 su[14666]: Successful su for pihole by root
Dez 10 11:46:04 ras3 su[14666]: + ??? root:pihole
Dez 10 11:46:04 ras3 su[14666]: pam_unix(su:session): session opened for user pihole by (uid=0)
Dez 10 11:46:20 ras3 systemd[1]: Started LSB: pihole-FTL daemon.

```

**Comment by Mcat12** [comment]
Is `pihole-FTL` in the PATH for `root`? The active (exited) just means that FTL started and successfully daemonized.

**Comment by dracenmarx** [comment]
Yes

```
root@ras3:~$ echo $PATH
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
root@ras3:~$ whereis pihole-FTL
pihole-FTL: /usr/bin/pihole-FTL
root@ras3:~$ pihole-FTL -v
vDev-437af07
```

**Comment by adamsewell** [comment]
I'm having a similar issue, the pihole-FTL will fail upon me logging into the dashboard, it will display the first stats but then changes to "lost connection". http://d.pr/i/7ajSWH/5KPL9doG 

It's almost like something is causing it to fail when stats are being pulled.

**Comment by Mcat12** [comment]
@adamsewell It sounds like your issue is different. Please open a new issue.

**Comment by Mcat12** [comment]
@dracenmarx What is the output of `killall -V` and `killall -h`?

**Comment by dracenmarx** [comment]
Note, in re @adamsewell: I am using pihole without web interface and without logging, so maybe the issue is different.

@mcat12 The output is:

$ killall -h
Aufruf: killall [-Z CONTEXT] [-u USER] [ -eIgiqrvw ] [-SIGNAL] NAME...
       killall -l, --list
       killall -V, --version

  -e,--exact          verlange genaue �ereinstimmung f�r sehr lange Namen
  -I,--ignore-case    Gro� und Kleinschreibung nicht beachten
  -g,--process-group  breche Prozessgruppe statt Einzelprozess ab
  -y,--younger-than   Prozesse j�nger als ZEIT abbrechen
  -o,--older-than     Prozesse �ter als ZEIT abbrechen
  -i,--interactive    verlange vor Abbruch Best�igung des Benutzers
  -l,--list           liste alle bekannten Signalnamen auf
  -q,--quiet          keine Warnungen und Fehler ausgeben
  -r,--regexp         NAME ist ein erweiteter regul�er Ausdruck
  -s,--signal SIGNAL  sende benutzerdefiniertes Signal anstatt SIGTERM
  -u,--user USER      nur Prozesse von angegebenem Benutzer abbrechen
  -v,--verbose        berichte, falls das Signal erfolgreich gesendet wurde
  -V,--version        zeige Version
  -w,--wait           warte auf das Ende der Prozesse
  -Z,--context REGEXP breche nur Prozesse ab, die einen Kontext haben
                        (muss vor anderen Argumenten stehen)

$ killall -V
killall (PSmisc) 22.21
Copyright (C) 1993-2012 Werner Almesberger und Craig Small

F�r PSmisc gibt es KEINERLEI GARANTIE.
Es ist freie Software und Sie d�rfen sie gern gem� den Bedingungen
der GNU General Public License (GPL) weiter vertreiben.
Zus�zliche Informationen dazu finden Sie in der Datei namens COPYING.

**Comment by Mcat12** [comment]
What is the output of running this?
```
sudo killall -s SIGHUP pihole-FTL
```

I did get the process not found message when I didn't use `sudo`, but with it running under root it executed fine.

**Comment by dracenmarx** [comment]
Same problem:

[23:53]pi@ras3:~$ sudo killall -s SIGHUP pihole-FTL
pihole-FTL: Kein Prozess gefunden
[23:53]pi@ras3:~$ killall -s SIGHUP pihole-FTL
pihole-FTL: Kein Prozess gefunden
[23:53]pi@ras3:~$ sudo su
[23:53]root@ras3:/home/pi$ sudo killall -s SIGHUP pihole-FTL
pihole-FTL: Kein Prozess gefunden
[23:53]root@ras3:/home/pi$ killall -s SIGHUP pihole-FTL
pihole-FTL: Kein Prozess gefunden

Are you sure that killall is the problem? because when I run "ps aux | grep pihole",  nothing is listed. so I think it is not running ; maybe it crashed somehow shortly after being demonized?

**Comment by Mcat12** [comment]
Does the web interface show any stats? Run `pihole -d` for a new debug token.

**Comment by dracenmarx** [comment]
My debug token is: pw6glxi4gd
Like I said before, I do not use logs or web interface.
(I use Apache on port 80, so I can't use pihole's webserver)

**Comment by Mcat12** [comment]
Sorry, I forgot that you didn't use the web interface. Since you don't use logs, then I would assume FTL would not have anything to do and might actually crash when trying to access the log file. What are the contents of `/var/log/pihole-FTL.log`?

@DL6ER this might be an unforeseen issue when logging is disabled?

**Comment by DL6ER** [comment]
> this might be an unforeseen issue when logging is disabled?

No, actually that is not unforeseen. If there is nothing for `FTL` to be done (as it is *only* working with the logs), there is also no reason for it to stay around, is it?

**Comment by dracenmarx** [comment]
Ok, I think I have found the reason why the FTL process is not running:

```
root@ras3:~$ cat pihole-FTL.log
[2017-12-16 01:21:49.137] WARN:  Opening of /var/log/pihole.log failed!
[2017-12-16 01:21:49.137]        Make sure it exists and is readable by user pihole
       Will try again in 15 seconds.
[2017-12-16 01:22:04.147] FATAL: Opening of /var/log/pihole.log failed permanently!


root@ras3:~$ ls /var/log/pihole*
-rw-r--r-- 1 root    pihole 998K Dez 14 14:27 /var/log/pihole_debug.log
-rw-r--r-- 1 root    root   997K Dez 14 14:27 /var/log/pihole_debug-sanitized.log
-rw-r--r-- 1 pihole  pihole  277 Dez 16 01:22 /var/log/pihole-FTL.log
-rw-r--r-- 1 pihole  pihole  277 Dez 16 01:21 /var/log/pihole-FTL.log.1
-rw-r--r-- 1 pihole  pihole  189 Dez 10 00:00 /var/log/pihole-FTL.log.2.gz
-rw-r----- 1 dnsmasq root      0 Dez 16 01:21 /var/log/pihole.log
-rw-r----- 1 dnsmasq root    16K Dez 16 01:21 /var/log/pihole.log.1
-rw-r----- 1 dnsmasq root   2,0K Dez 10 00:00 /var/log/pihole.log.2.gz
```

The pihole.log files are owned by the user dnsmasq. But I did not change anything at all. Everything is out of the box, except that /var/log/ is a RAM-disk, so it is flushed on every reboot. So, pihole.log will automatically re-created with that dnsmasq user. Is that a bug? Because I expect that pihole.log should be owned by pihole:pihole and not dnsmasq:root

**Comment by Mcat12** [comment]
The issue is that `/var/log` is on a RAM disk, so the original permissions the installer applied were wiped later on. There's a fix for this coming in the next release.

**Comment by DL6ER** [comment]
Which version of `pihole-FTL` are you running?

Try
```
pihole-FTL version
```

Concerning file permissions/ownership:
```
-rw-r----- 1 dnsmasq root      0 Dez 16 01:21 /var/log/pihole.log
```
is a problem because only user `dnsmasq` and group `root` can read the file. It is correct that they are the owners. However, all other users can't read anything.
The fix for this is to change the permissions to allow `read` for all, like
```
-rw-r--r-- 1 dnsmasq root      0 Dez 16 01:21 /var/log/pihole.log
```

If the system is not on a RAM disk, then this change has to be done only once. Not sure how to set this up on a RAM disk setup, because the permissions would have to get corrected *after* `dnsmasq` started and before `pihole-FTL` exists due to not being able to read the log file.

**Comment by dracenmarx** [comment]
Thank you for your hint; yes, applying the read-rights solves the problem. pihole-FTL stays running, and during pihole setup there is no warning.

So, I think there could be a fix to support RAM disks (and/or errornous log rotation).

**Comment by DL6ER** [comment]
So far `FTL` terminates when it cannot read its only data source. We could leave it running but it would just sit around not being able to deliver any data. Do you think this would be any better?

Note that `FTL` cannot change file permissions on its own as it is designed with properly configured permissions which e.g. allow it to only access files which are accessible for user `pihole`. It has absolutely zero `sudo` powers.

**Comment by dracenmarx** [comment]
In this case I think it is correct that FTL exists.
But I think the process that initially creates the log file should somehow apply the correct permissions. I do not understand why dnsmasq generates a file which should belong to pihole (or at least is named pihole).
I do not think it is good that accidently deleting a log file (or using a RAM disk, or having an errornous log rotation) will defunct FTL.

**Comment by DL6ER** [comment]
> the process that initially creates the log file should somehow apply the correct permissions

This is `dnsmasq`. It does that such that only itself and the user `root` (or any user with `sudo`-fu) can read the content of this file. Imagine a larger DHCP/DNS-server in a company where users have `ssh`-logins. If the `dnsmasq` would be readable by all, that would be a quite severe privacy issue.

> I do not understand why dnsmasq generates a file which should belong to pihole (or at least is named pihole).

We tell it to just name the file like this (via `dnsmasq`'s config file). Not more.

> I do not think it is good that accidently deleting a log file (or using a RAM disk, or having an errornous log rotation) will defunct FTL permanently.

Do you have any suggestions in the light of what I wrote above on how to improve on the situation?

**Comment by dracenmarx** [comment]
> Do you have any suggestions in the light of what I wrote above on how to improve on the situation?

Yes, I would create a group dnsmasq, and let pihole.log be owned by dnsmasq:dnsmasq . In the group we can have dnsmasq and pihole. So, both can read the file.
A lot of programs I know use this technique of creating an individual group, so that other processes can access data is required.

Mcat12 wrote
> The issue is that /var/log is on a RAM disk, so the original permissions the installer applied were wiped later on. There's a fix for this coming in the next release.

What is the fix he mentioned?

**Comment by Mcat12** [comment]
This PR applies the correct permissions: #1817

**Comment by jacobsalmela** [comment]
#1817 was merged and included in [v3.2.1](https://github.com/pi-hole/pi-hole/releases/tag/v3.2.1)

**Comment by Setaran** [comment]
I'm having this same issue on the latest version of pi-hole.
After installation I get `pihole-FTL: No process found`
And:
```
which pihole-FTL
/usr/bin/pihole-FTL
```
Also:
```
file /usr/bin/pihole-FTL
/usr/bin/pihole-FTL: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/l, for GNU/Linux 2.6.32, BuildID[sha1]=7e03376068bd0b14ff9641a026456c06ada57f41, with debug_info, not stripped
```
Please help!
Thanks for your time.

**Comment by dschaper** [comment]
Please open a new issue and follow the template. This issue was from the end of 2017 and a lot has changed in that time.
